### PR TITLE
[Move] Partially Implement Instruct

### PR DESCRIPTION
**Pull request by Bertie690**
[Locales PR](https://github.com/pagefaultgames/pokerogue-locales/pull/47) (merged)

## What are the changes the user will see?
Instruct will actually work and calls the target's last used move correctly.
It currently does not play nice with Gigaton Hammer, Blood Moon and Torment, but those are related to the janky implementation of those moves currently. #4289 
Also pledge jank happens currently.
Also the enemy cannot really use it atm but that's a topic for the upcoming AI reworks!!!

## Why am I making these changes?
Because I want funny klefki prankster instructing power trick shuckle rock slide. 

## What are the changes from a developer perspective?
Added `RepeatMoveAttr` for moves that force targets to repeat moves themselves. (This is distinct from `CopyMoveAttr` which has the *user* copy a target's previous move)

### Screenshots/Videos
Instruct repeating an ally's rock slide & subsequently failing on enemy Unown due to it not having made moves yet:
https://github.com/user-attachments/assets/d4ee43d5-d324-4f5e-824d-76aac5e26ea1
It sorta works on pledges... a bit
https://github.com/user-attachments/assets/60490c80-57bd-43cd-8d0b-66ea511803bb
Instruct fails when used on Instruct (Full list of moves on Bulbapedia):
https://github.com/user-attachments/assets/b9a22fa2-ab67-4d1d-bd6b-81b855fa545a

## How to test the changes?
Run `npm run test instruct`

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/move.ts [hunk 1]
```diff
@@ -6623,6 +6623,124 @@ export class CopyMoveAttr extends OverrideMoveEffectAttr {
   }
 }
 
+/**
+ * Attribute used for moves that causes the target to repeat their last used move.4
+ *
+ * Used for [Instruct](https://bulbapedia.bulbagarden.net/wiki/After_You_(move)).
```
Comment (torranx) [inline src/data/move.ts:6629]: link says After_You?

**Review by torranx — COMMENTED** [review]
stopped reviewing after seeing a lot of TODOs, maybe place this in draft for now if its not ready for review - or do you intend them to be in a separate PR?

**Comment by torranx** [comment]
also is there a reason why this is not using `MoveRestrictionBattlerTag`? How does this implementation interact with moves like Disable and Throat Chop?

**Comment by Bertie690** [comment]
> stopped reviewing after seeing a lot of TODOs, maybe place this in draft for now if its not ready for review - or do you intend them to be in a separate PR?

My thought is that those fixes can probably go in a separate PR. This is _partially_ implementing it after all!

> also is there a reason why this is not using `MoveRestrictionBattlerTag`? How does this implementation interact with moves like Disable and Throat Chop?

Testing it again now, but the intended behavior is that Instruct itself will succeed but the called move will fail.

File: src/test/moves/instruct.test.ts [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+  it("should try to repeat enemy's disabled move, but fail", async () => {
+    game.override.enemyMoveset([ Moves.SONIC_BOOM, Moves.PROTECT, Moves.SUBSTITUTE, Moves.HYPER_BEAM ]);
+    game.override.moveset([ Moves.INSTRUCT, Moves.SONIC_BOOM, Moves.DISABLE, Moves.SPLASH ]);
+    await game.classicMode.startBattle([ Species.AMOONGUSS, Species.DROWZEE ]);
+
+    game.move.select(Moves.INSTRUCT, BattlerIndex.PLAYER, BattlerIndex.ENEMY);
+    game.move.select(Moves.DISABLE, BattlerIndex.PLAYER_2, BattlerIndex.ENEMY);
+    await game.forceEnemyMove(Moves.SONIC_BOOM, BattlerIndex.PLAYER);
+    await game.setTurnOrder([ BattlerIndex.ENEMY, BattlerIndex.PLAYER_2, BattlerIndex.PLAYER ]);
```
[the remark this answers is not in the document]
Comment (Bertie690) [reply src/test/moves/instruct.test.ts:174]: yes - they move first, then we disable their move and instruct it afterwards.

File: src/test/moves/instruct.test.ts [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    game.move.select(Moves.INSTRUCT, BattlerIndex.PLAYER, BattlerIndex.ENEMY);
+    await game.forceEnemyMove(Moves.SONIC_BOOM, BattlerIndex.PLAYER);
+    await game.setTurnOrder([ BattlerIndex.PLAYER, BattlerIndex.ENEMY ]);
+    await game.phaseInterceptor.to("TurnEndPhase", false);
+
+    // should fail to execute
+    expect(game.scene.getPlayerPokemon()!.getLastXMoves()[0].result).toBe(MoveResult.FAIL);
+  });
+
+  it("should try to repeat enemy's disabled move, but fail", async () => {
```
[the remark this answers is not in the document]
Comment (Bertie690) [reply src/test/moves/instruct.test.ts:166]: ok mb - i can change the description quickly.

**Comment by torranx** [comment]
> My thought is that those fixes can probably go in a separate PR. This is _partially_ implementing it after all!

no problem 👍 

> Testing it again now, but the intended behavior is that Instruct itself will succeed but the called move will fail.
Torment is a special case: it should prevent move selection but not move execution.

ok so I've misunderstood the move, re-read the move description and watched some clips - `MoveRestrictionBattlerTag` is not the way to go yes: here's a detailed [yt video](https://www.youtube.com/watch?v=Pj0xfNEqeqM) on the move's mechanics.

**Review by DayKev — COMMENTED** [review]
Found an issue where because Instruct was causing the Pokémon to use 2 moves in a single turn, the `firstHit` value was false in `MoveEffectPhase` for the second move when it shouldn't be. I fixed it for when the Instruct user moves after the target in the same turn, but the issue will still happen if the Instruct user moves first. Still need to look into that scenario.

Also fixed up the tests.

Since there were too many changes to reasonably use GitHub's suggestions feature, I made a PR instead: https://github.com/Bertie690/pokerogue/pull/1

**Comment by Bertie690** [comment]
> Since there were too many changes to reasonably use GitHub's suggestions feature, I made a PR instead: https://github.com/Bertie690/pokerogue/pull/1

Cool. When do you want me to merge it? 
(I'd presume by your earlier part of the message that the bug isn't 100% fixed yet)

**Comment by DayKev** [comment]
You can go ahead and merge it now, and I'll take a look at making a follow-up later today.

**Comment by DayKev** [comment]
I think I found a clean solution to the multi-hits issue, updated the PR since it wasn't merged yet.

File: src/data/move.ts [hunk 4]
```diff
@@ -9812,7 +9931,7 @@ export function initMoves() {
       .attr(StatStageChangeAttr, [ Stat.ATK ], -1),
     new StatusMove(Moves.INSTRUCT, Type.PSYCHIC, -1, 15, -1, 0, 7)
       .ignoresSubstitute()
-      .unimplemented(),
+      .attr(RepeatMoveAttr),
```
Comment (innerthunder) [inline src/data/move.ts:9934]:
```suggestion
      .attr(RepeatMoveAttr)
      .edgeCase(), // incorrect interactions with Gigaton Hammer, Blood Moon, Torment, and the Pledges
```
Adding an edge case for the buggy interactions. I figure the "problem moves" are specific enough to not warrant a `.partial()` marker.

**Review by innerthunder — DISMISSED** [review]

**Review by torranx — DISMISSED** [review]

**Review by DayKev — APPROVED** [review]

**Comment by MokaStitcher** [comment]
@Bertie690 what's the status on the locale PR for this one? Seems like it's still marked as draft

**Comment by Bertie690** [comment]
🤷
AFAIK, it seems the non-English locale keys haven't been added yet.

**Comment by MokaStitcher** [comment]
> AFAIK, it seems the non-English locale keys haven't been added yet.

Well yes, since the PR for the English key is still in draft there is no chance translations have been done. Why is it in draft is my question, is it not ready and needed for thttps://github.com/pagefaultgames/pokerogue-locales/pull/47

**Comment by MokaStitcher** [comment]
Sorry for the accidental closure of the PR 💀

**Review by frutescens — APPROVED** [review]

**Pull request merged**